internal/extension/smartagentextension/config.go

```diff
@@ -26,50 +26,62 @@ import (
 	"gopkg.in/yaml.v2"
 )
 
-// SmartAgentConfigProvider exposes config fields to other packages.
-// This is needed since fields such  as bundleDir and collectdConfig
-// are mapped to camel case fields in the config and hence are not exposed.
+// SmartAgentConfigProvider exposes global saconfig.Config to other components
 type SmartAgentConfigProvider interface {
-	BundleDir() string
-	CollectdConfig() *saconfig.CollectdConfig
+	SmartAgentConfig() *saconfig.Config
 }
 
 var _ SmartAgentConfigProvider = (*Config)(nil)
 
 type Config struct {
 	config.ExtensionSettings `mapstructure:",squash"`
-	bundleDir                string
-	collectdConfig           saconfig.CollectdConfig
+	// Agent uses yaml, which mapstructure doesn't support.
+	// Custom unmarshaller required for yaml and SFx defaults usage.
+	saconfig.Config `mapstructure:"-,squash"`
 }
 
-func (c Config) BundleDir() string {
-	return c.bundleDir
-}
-
-func (c Config) CollectdConfig() *saconfig.CollectdConfig {
-	return &c.collectdConfig
+func (c Config) SmartAgentConfig() *saconfig.Config {
+	return &c.Config
 }
 
 func customUnmarshaller(componentViperSection *viper.Viper, intoCfg interface{}) error {
 	allSettings := componentViperSection.AllSettings()
 	extensionCfg := intoCfg.(*Config)
 
-	if bundleDir, ok := allSettings["bundledir"]; ok {
-		extensionCfg.bundleDir = fmt.Sprintf("%s", bundleDir)
-		delete(allSettings, "bundledir")
+	config, err := smartAgentConfigFromSettingsMap(allSettings)
+	if err != nil {
+		return err
+	}
+
+	if config.BundleDir == "" {
+		config.BundleDir = extensionCfg.Config.BundleDir
+	}
+	config.Collectd.BundleDir = config.BundleDir
+
+	extensionCfg.Config = *config
+	return nil
+}
+
+func smartAgentConfigFromSettingsMap(settings map[string]interface{}) (*saconfig.Config, error) {
+	var config saconfig.Config
+	yamlTags := yamlTagsFromStruct(reflect.TypeOf(config))
```

Comment:
Do you know if the SA top level config requires the recursive version of this method available in the receiver code? For the collectd config it wasn't required. I believe we discussed moving this method to a utils package. If you require the recursive version here, I can open a PR moving it since I said I'll do it but haven't :)

Reply:
It would probably be preferable to what I have here.  Will do so in this PR if that's ok w/ you.

Reply:
That's fine as well.

Reply:
@asuresh4 after thinking about this I think the recursive version is unnecessary since the case correction only occurs at the root level, or here manually for the additional collectd updates.  In the future we will probable want a more robust helper to coerce arbitrary maps of maps to desired cases but that's not needed here imo.